src/main/java/com/rtb/projectmanagementtool/taskblocker/TaskBlockerController.java

```diff
@@ -2,89 +2,138 @@
+import com.google.appengine.api.memcache.MemcacheService;
 import com.rtb.projectmanagementtool.task.*;
 import com.rtb.projectmanagementtool.task.TaskData.Status;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.LinkedList;
-import java.util.Map;
-import java.util.stream.Collectors;
+import org.apache.commons.lang3.SerializationUtils;
 
 /** Class controlling the TaskBlockerData object. */
 public final class TaskBlockerController {
   private DatastoreService datastore;
+  private MemcacheService cache;
   private TaskController taskController;
   private Filter NO_FILTER = null;
 
   public TaskBlockerController(DatastoreService datastore, TaskController taskController) {
     this.datastore = datastore;
+    this.cache = null;
+    this.taskController = taskController;
+  }
+
+  public TaskBlockerController(
+      DatastoreService datastore, MemcacheService cache, TaskController taskController) {
+    this.datastore = datastore;
+    this.cache = cache;
     this.taskController = taskController;
   }
 
   // Add methods
 
-  public void addTaskBlocker(long taskID, long blockerID) {
+  public void addTaskBlocker(long taskID, long blockerID) throws TaskBlockerException {
     addTaskBlocker(new TaskBlockerData(taskID, blockerID));
   }
 
-  public void addTaskBlocker(TaskBlockerData taskBlocker) {
+  public void addTaskBlocker(TaskBlockerData taskBlocker) throws TaskBlockerException {
     // Ensure the tasks exist
     ArrayList<TaskData> tasks =
         taskController.getTasksByIDs(
             new ArrayList<>(Arrays.asList(taskBlocker.getTaskID(), taskBlocker.getBlockerID())));
     if (tasks.size() < 2) {
-      System.out.println("Cannot find tasks with provided taskID or blockerID");
-      return;
+      throw new TaskBlockerException("Cannot find tasks with provided taskID or blockerID");
     }
     // Ensure the blocked task isn't already set to COMPLETE
     if (tasks.get(0).getStatus() == Status.COMPLETE
         || tasks.get(1).getStatus() == Status.COMPLETE) {
-      System.out.println("One or more tasks are already set to COMPLETE.");
-      return;
+      throw new TaskBlockerException("One or more tasks are already set to COMPLETE");
+    }
+    // Ensure the blocked task isn't a subtask of the blocker
+    if (taskController.isSubtask(taskBlocker.getBlockerID(), taskBlocker.getTaskID())) {
+      throw new TaskBlockerException("A task cannot block its subtask");
     }
     // Ensure a cycle wouldn't be created if the TaskBlocker is added
     if (containsPath(taskBlocker.getBlockerID(), taskBlocker.getTaskID())) {
-      System.out.println("Cannot block a task if it would create a cycle.");
-      return;
+      throw new TaskBlockerException("Cannot block a task if it would create a cycle");
     }
     // Add the TaskBlocker
     taskBlocker.setTaskBlockerID(datastore.put(taskBlocker.toEntity()).getId());
   }
 
   private boolean containsPath(long start, long end) {
```

Comment:
This can be a method of the graph class, since finding a path could be a valid operation for the graph itself.  One advantage of moving this in the graph: if you switch graph implementations for example from HashMaps to actual Node classes, you wouldn't have to change any of the files that requires checking for paths.